src/client/dfuse/pil4dfs/hook.c

```diff
@@ -172,77 +179,88 @@ read_map_file(char **buf, const char str_target[], char **pos)
 static void
 determine_lib_path(void)
 {
-	int   i, rc;
 	char *read_buff_map = NULL;
-	char *pPos = NULL, *pStart = NULL, *pEnd = NULL;
-	char  lib_ver_str[32], lib_dir_str[256];
+	char *pos, *start, *end;
+
+	read_map_file(&read_buff_map);
 
-	read_map_file(&read_buff_map, "/ld-2.", &pPos);
-	if (pPos == NULL) {
-		free(read_buff_map);
-		found_libc = 0;
+	pos = strstr(read_buff_map, "/lib64/ld-linux");
+	if (pos == NULL)
+		/* try a different format */
+		pos = strstr(read_buff_map, "/lib64/ld-2");
+	if (pos == NULL) {
 		printf("Warning: Failed to find ld.so.\n");
-		return;
+		goto err;
 	}
-	for (i = 0; i < 16; i++) {
-		if (strncmp(pPos + i, ".so", 3) == 0) {
-			pEnd = pPos + i + 3;
-			break;
-		}
+	get_string_pos(pos, &start, &end);
+	if ((end - start + 1) >= PATH_MAX) {
+		printf("Warning: path of path_ld is too long.\n");
+		goto err;
 	}
-	if (pEnd == NULL) {
-		printf("Fail to determine the ending position of libc path.\nQuit\n");
-		exit(1);
+	path_ld = malloc(end - start + 1);
+	if (path_ld == NULL) {
+		printf("Warning: Failed to allocate memory for path_ld.\n");
+		goto err;
 	}
-	for (i = 0; i < 100; i++) {
-		if (strncmp(pPos - i, "  /", 3) == 0) {
-			pStart = pPos - i + 2;
-			break;
-		}
+	memcpy(path_ld, start, end - start);
+	path_ld[end - start] = 0;
+
+	pos = strstr(read_buff_map, "/lib64/libc.so");
+	if (pos == NULL)
+		/* try a different format */
+		pos = strstr(read_buff_map, "/lib64/libc-2");
+	if (pos == NULL) {
+		printf("Warning: Failed to find libc.so.\n");
+		goto err;
 	}
-	if (pStart == NULL) {
-		printf("Fail to determine the starting position of libc path.\nQuit\n");
-		exit(1);
+	get_string_pos(pos, &start, &end);
+	if ((end - start + 1) >= PATH_MAX) {
+		printf("The length of path_libc is too long!\nQuit\n");
+		goto err;
 	}
-
-	if ((long int)(pEnd - pStart) >= PATH_MAX) {
-		printf("The length of path_ld is too long!\nQuit\n");
-		exit(1);
+	path_libc = malloc(end - start + 1);
+	if (path_libc == NULL) {
+		printf("Warning: Failed to allocate memory for path_libc.\n");
+		goto err;
 	}
-	path_ld = malloc((int)(pEnd - pStart) + 1);
-	if (path_ld == NULL) {
-		printf("Failed to allocate memory for path_ld.\nQuit\n");
-		exit(1);
+	_Pragma("GCC diagnostic push")
+	_Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
+	_Pragma("GCC diagnostic ignored \"-Wstringop-overflow\"")
+	_Pragma("GCC diagnostic ignored \"-Wrestrict\"")
+	memcpy(path_libc, start, end - start);
+	path_libc[end - start] = 0;
+	_Pragma("GCC diagnostic pop")
```

Comment:
Same comment as above, why is strndup not adequate here?  Please don't disable compiler warnings in the code unless you absolutely have to.

Reply:
D_STRNDUP is used.